File: 1356.sort-integers-by-the-number-of-1-bits.cpp
/*
 * @lc app=leetcode id=1356 lang=cpp
 *
 * [1356] Sort Integers by The Number of 1 Bits
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  vector<int> sortByBits(vector<int>& arr) {
    std::unordered_map<int, int> counts;
    for (const auto num : arr) {
      int count = 0;
      int tmp = num;
      while (tmp) {
        count += tmp % 2;
        tmp /= 2;
      }
      counts[num] = count;
    }
    std::sort(std::begin(arr), std::end(arr), [&counts](int lhs, int rhs) {
      int lbits = counts[lhs];
      int rbits = counts[rhs];
      if (lbits < rbits) {
        return true;
      }
      if (lbits > rbits) {
        return false;
      }

      return lhs < rhs;
    });
    return arr;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<int> arr, exp;
  arr = {0, 1, 2, 3, 4, 5, 6, 7, 8};
  exp = {0, 1, 2, 4, 8, 3, 5, 6, 7};
  ASSERT_EQ(s.sortByBits(arr), exp);

  arr = {1024, 512, 256, 128, 64, 32, 16, 8, 4, 2, 1};
  exp = {1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024};
  ASSERT_EQ(s.sortByBits(arr), exp);

  arr = {10000, 10000};
  exp = {10000, 10000};
  ASSERT_EQ(s.sortByBits(arr), exp);

  arr = {2, 3, 5, 7, 11, 13, 17, 19};
  exp = {2, 3, 5, 17, 7, 11, 13, 19};
  ASSERT_EQ(s.sortByBits(arr), exp);

  arr = {10, 100, 1000, 10000};
  exp = {10, 100, 10000, 1000};
  ASSERT_EQ(s.sortByBits(arr), exp);
  return 0;
}

File: 1356.sort-integers-by-the-number-of-1-bits.rs
/*
 * @lc app=leetcode id=1356 lang=rust
 *
 * [1356] Sort Integers by The Number of 1 Bits
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn sort_by_bits(mut arr: Vec<i32>) -> Vec<i32> {
        arr.sort_by(|a, b| {
            let ones_a = a.count_ones();
            let ones_b = b.count_ones();
            if ones_a == ones_b {
                a.partial_cmp(b).unwrap()
            } else {
                ones_a.partial_cmp(&ones_b).unwrap()
            }
        });
        arr
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(
        Solution::sort_by_bits(vec![0, 1, 2, 3, 4, 5, 6, 7, 8]),
        vec![0, 1, 2, 4, 8, 3, 5, 6, 7]
    );
}

#[test]
fn test2() {
    assert_eq!(
        Solution::sort_by_bits(vec![1024, 512, 256, 128, 64, 32, 16, 8, 4, 2, 1]),
        vec![1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]
    );
}
